reason-parser/src/reason_pprint_ast.ml

```diff
@@ -3375,8 +3399,8 @@ class printer  ()= object(self:'self)
     | PotentiallyLowPrecedence itm -> formatPrecedence ~loc:reducesAfterRight.pexp_loc itm
     | Simple itm -> itm
 
-  method ensureExpression expr ~reducesOnToken =
-    match self#unparseExprRecurse expr with
+  method ensureExpression ?(infixChain=false) ~reducesOnToken expr =
+    match self#unparseExprRecurse ~infixChain expr with
     | SpecificInfixPrecedence ({reducePrecedence; shiftPrecedence}, leftRecurse) ->
```

Comment:
This would likely be easier to implement this change, as well as all the other customizations you mentioned, if the result of unparsing returned an intermediate representation of the *concrete* AST. That is, it would return a tree where all the precedence has already been resolved. Imagine if you had something like this:

```
InfixApplicationConcrete (
  "|>",
  InfixApplicationConcrete ( "|>", SimpleConcrete itm),
  SimpleConcrete itm
)
```

Where printing is trivial, and all parenthesis/precedence has already been handled.
Then you can easily do things like
```
computeConsecutiveLeftInfixChains (concreteAst)
```

which would return a convenient list for you to format in a specific way.

This division separates all the technical details of precedence and *correctness* from the opinionated *styling* of formatting.

Maybe the current `FunctionApplication`/`Simple`/`SpecificInfixPrecedence` could *become* that Concrete AST I mentioned? It's pretty close to it.